File: applications/petascope/src/main/java/petascope/util/ListUtil.java
/*
 * This file is part of rasdaman community.
 *
 * Rasdaman community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rasdaman community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU  General Public License for more details.
 *
 * You should have received a copy of the GNU  General Public License
 * along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2003 - 2014 Peter Baumann / rasdaman GmbH.
 *
 * For more information please see <http://www.rasdaman.org>
 * or contact Peter Baumann via <baumann@rasdaman.com>.
 */
/*
 * JOMDoc - A Java library for OMDoc documents (http://omdoc.org/jomdoc).
 *
 * Original author    Normen Müller <n.mueller@jacobs-university.de>
 * Web                http://kwarc.info/nmueller/
 * Created            Oct 17, 2007
 * Filename           $Id: ListUtil.java 1976 2010-07-31 12:07:20Z dmisev $
 * Revision           $Revision: 1976 $
 *
 * Last modified on   $Date:2007-10-25 18:50:01 +0200 (Thu, 25 Oct 2007) $
 *               by   $Author:nmueller $
 *
 * Copyright (C) 2007,2008 the KWARC group (http://kwarc.info)
 * Licensed under the GNU  Public License v3 (GPL3).
 * For other licensing contact Michael Kohlhase <m.kohlhase@jacobs-university.de>
 */
package petascope.util;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import nu.xom.Node;

/**
 * List utilities.
 *
 * @author Normen M&#xFC;ller&nbsp;&#60;n.mueller@jacobs-university.de&#62;
 * @author <a href="mailto:d.misev@jacobs-university.de">Dimitar Misev</a>
 */
public class ListUtil {

    // TODO comment
    public static final <T> List<T> reverse(List<T> l) {
        List<T> nl = new LinkedList<T>(l);
        Collections.reverse(nl);
        return nl;
    }

    /**
     * @return sublist(l, n, m) --> [x_n,...,x_m] if 0 <= n < m < size(l) else empty_list
     */
    public static <T> List<T> sublist(List<T> l, int fromIndex, int toIndex) {
        if (fromIndex > toIndex) {
            return Collections.<T>emptyList();
        }
        return l.subList(fromIndex, toIndex + 1);
    }

    /**
     * Converts a collection to a string, separating the elements by ","
     *
     * @param l The StringList
     * @param c The delimiter
     * @return A String of the ListElements separated by c.
     */
    public static <T> String ltos(Collection<T> l, String c) {
        String s = "";
        for (Iterator<T> iter = l.iterator(); iter.hasNext();) {
            if (!s.equalsIgnoreCase("")) {
                s = s + c + iter.next().toString();
            } else {
                s = s + iter.next().toString();
            }
        }
        return s;
    }

    /**
     * Converts a collection  to a string
     *
     * @param l the element list
     * @return A String of the ListElements separated by new lines
     */
    public static <T> String ltos(Collection<T> l) {
        String s = "";
        for (Iterator<T> iter = l.iterator(); iter.hasNext();) {
            Object o = iter.next();
            String tmp = null;
            if (o instanceof Node) {
                tmp = ((Node) o).toXML();
            } else if (o instanceof Collection) {
                tmp = ltos((Collection) o);
            } else {
                tmp = o.toString();
            }
            if (!s.equals("")) {
                s += "\n";
                s += tmp;
            } else {
                s += tmp;
            }
        }
        return s;
    }

    /**
     * Converts a string to a list
     *
     * @param s The StringList
     * @param c The delimiter
     * @return A String of the ListElements separated by c.
     */
    public static List<String> stol(String s, String c) {
        List<String> l = new LinkedList<String>();
        if (s == null) {
            return l;
        }
        s = s.replaceAll(" *, *", c);
        String[] sl = s.split(c, s.length());
        for (int i = 0; i < sl.length; i++) {
            l.add(sl[i]);
        }
        return l;
    }

    /* head(e_1,...,e_n) --> e_1 */
    public static <T> T head(List<T> l) {
        if (l == null || l.isEmpty()) {
            return null;
        } else {
            return l.get(0);
        }
    }

    /* head(e_1,...,e_n) --> (e_2,...,e_n) */
    public static <T> List<T> tail(List<T> l) {
        if (l == null) {
            return null;
        }
        if (l.isEmpty() || l.size() == 1) {
            return Collections.<T>emptyList();
        }
        return ListUtil.sublist(l, 1, l.size() - 1);
    }

    /**
     * con (e, [x_1,...x_n]) --> [e,x_1,...x_n]
     */
    public static <T> List<T> cons(T e, List<T> l) {
        return concat(Collections.singletonList(e), l);
    }

    /**
     * [x_1,...x_n]@[y_1,...y_m] --> [x_1,...x_n,y_1,...,y_m]
     */
    public static <T> List<T> concat(List<T> l, List<T> r) {
        List<T> res = new ArrayList<T>();
        for (T a : l) {
            res.add(a);
        }
        for (T a : r) {
            res.add(a);
        }
        return res;
    }

    public static <T> List<T> toList(T... e) {
         List<T> ret = new ArrayList<T>();
         if (e != null) {
             for (T o : e) {
                ret.add(o);
            }
         }
         return ret;
    }

    /**
     * New iteration [x_1,...x_m] --> [x_1,...x_m, n]
     */
    public static final List<Integer> add(List<Integer> l, int n) {
        List<Integer> nl = new LinkedList<Integer>(l);
        nl.add(n);
        return nl;
    }

    /**
     * Goto next iteration [x_1,...x_m] --> [x_1,...x_m + 1]
     */
    public static final List<Integer> inc(List<Integer> l) {
        List<Integer> nl = new LinkedList<Integer>(l);
        if (!nl.isEmpty()) {
            nl.add(nl.remove(nl.size() - 1).intValue() + 1);
        }
        return nl;
    }

    /**
     * @param <T> return type
     * @param <E> input type
     * @param l input list
     */
    public static <T, E> List<T> cast(List<E> l) {
        List<T> ret = new ArrayList<T>();
        for (Object object : l) {
            ret.add((T) object);
        }
        return ret;
    }

    public static <T> T[] append(T[] a, T... b) {
        T[] ret = (T[]) new Object[a.length + b.length];
        for (int i = 0; i < a.length; i++) {
            ret[i] = a[i];
        }
        for (int i = 0; i < b.length; i++) {
            ret[i + a.length] = b[i];
        }
        return ret;
    }

    // Returns the index of the last element lower than "el".
    // NOTE: it is assumed the array has been sorted and is non-decreasing.
    public static <T extends Number> int minIndex(List<T> a, T el) {

        if (el.getClass().equals(BigDecimal.class)) {
            for (int i=0; i < a.size(); i++) {
                if (((BigDecimal)a.get(i)).compareTo((BigDecimal)el) >= 0) {
                    return i;
                }
            }
        } else {
            for (int i=0; i < a.size(); i++) {
                if (a.get(i).doubleValue() >= el.doubleValue()) {
                    return i;
                }
            }
        }

        // "el" was outside bounds
        return a.size()-1;
    }


    /**
     * Returns the relative orders of the element in a numeric list.
     * @param <T>
     * @param list
     * @return
     */
    public static <T extends Number> List<Integer> relativeOrders(List<T> list) {
        List<Integer> relativeOrders = new ArrayList<Integer>(list.size());

        // sort the input list
        List<T> sortedList = new ArrayList<T>(list.size());
        sortedList.addAll(list);
        Collections.sort(sortedList, new NumericComparator());

        for (T el : list) {
            relativeOrders.add(sortedList.indexOf(el));
        }

        return relativeOrders;
    }

    /**
     * Print all elements in the list, separated by the specified Field Separator.
     * @param <T>
     * @param l    The list
     * @param FS   The field separator (e.g. comma for CSV outputs)
     * @return Readable listing of all elements in the list. Example: printList({a,b,c}, "-") --> "a-b-c"
     */
    public static <T> String printList(List<T> l, String FS) {
        String out = "";
        StringBuilder sb = new StringBuilder(out);
        if (l.size()>0) {
            boolean isBigDecimal = l.get(0).getClass().equals(BigDecimal.class);
            for (int i=0; i<l.size()-1; i++) {
                if (isBigDecimal) {
                    sb.append(BigDecimalUtil.stripDecimalZeros((BigDecimal)l.get(i)));
                } else {
                    sb.append(l.get(i));
                }
                sb.append(FS);
            }
            if (isBigDecimal) {
                    sb.append(BigDecimalUtil.stripDecimalZeros((BigDecimal)l.get(l.size()-1)));
                } else {
                    sb.append(l.get(l.size()-1));
                }
        }
        return sb.toString();
    }

    /**
     * Generic comparator for numbers.
     * @param <T>
     */
    private static class NumericComparator<T extends Number> implements Comparator<T> {

        @Override
        public int compare(T number1, T number2) {
            int out = 0; // n1=n2

            if (number1.getClass().equals(BigDecimal.class)) {
                out = ((BigDecimal)number1).compareTo((BigDecimal)number2);
            } else {
                if (number1.doubleValue() > number2.doubleValue()) {
                    out = 1;
                } else if (number1.doubleValue() < number2.doubleValue()) {
                    out = -1;
                }
            }

            return out;
        }
    }
}

File: applications/wcst_import/util/list_util.py
"""
 *
 * This file is part of rasdaman community.
 *
 * Rasdaman community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rasdaman community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU  General Public License for more details.
 *
 * You should have received a copy of the GNU  General Public License
 * along with rasdaman community.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2003 - 2020 Peter Baumann / rasdaman GmbH.
 *
 * For more information please see <http://www.rasdaman.org>
 * or contact Peter Baumann via <baumann@rasdaman.com>.
 *
"""
import decimal


def numpy_array_to_list_decimal(numpy_array):
    """
    Convert a numpy array to list of bigdecimal without losing precision.
    This is used for the cases of netcdf (or grib), e.g: ncdump shows the time axis with: 148654.084259259, 148656.99994213
    but, actually the list contains values in numpy array: 148654.08425925925, 148656.99994212962, 148658.08229166668
    and min/max (numpy_array) will return the value as ncdump which causes problem with coefficient in petascope
    (e.g: the first coefficient for irregular time axis is like 2.810E-7 not 0.0) and coverage will not be ingested.

    The reason is, for instance: "directPositions": "[float(x) * 24 * 3600 - 11644560000.0 for x in ${netcdf:variable:ansi}]"
    will return the whole numpy of array with precision so some operations to apply on input numpy_array like min(), max(),
    first(), last() *must* have same input list of decimal with same precision.

    :param array (numpy) numpy_array: array of values in numpy need to be convert to list of decimal without losing precision
    :rtype: list decimal
    """
    # NOTE: str(list(variable[:])) is *must* to keep all the precision of float values in numpy array
    output_list = list(map(decimal.Decimal, str(list(numpy_array))[1:-1].split(",")))

    return output_list


def numpy_array_to_string(numpy_array):
    """
    Convert a numpy array to string which represents the list of values with precision
    :param numpy_array: array of values in numpy
    :return: str: the representation string of numpy array
    """
    return str(list(numpy_array))


def to_list_string(input_list):
    """
    This is used to translate a list of decimal values to list string values as eval() cannot evaluate
    [Decimal(1), Decimal(2),...,Decimal(5)]
    :param input_list: list of decimal values
    :return: list string
    """
    output_list = [str(value) for value in input_list]
    return output_list


def sort_slices_by_datetime(slices, reverse=False):
    """
    Sort a list of slices (axes subsets from all analyzed files) by datetime
    e.g: we have a list of files: a_2005-01-01, b_2004-01-01, c_2003-01-01 and the datetime is extracted from file name
    Without the sort, the slices will be "2005-01-01", "2004-01-01", "2003-01-01" and cannot import in petascope
    as only can add the slice in top of a coverage for datetime axis.
    :param list input_list: list of slices
    :param bool reverse: sort list by reverse order
    :return: list of sorted slices by datetime axis
    """
    if len(slices) > 0:
        # Determine which axis contains datetime string
        time_axis_order = __get_time_axis_order(slices[0].axis_subsets)
        # Time axis exists in the list of axis subsets
        if time_axis_order != None:
            # sort the list by the date time axis order
            slices.sort(key=lambda slice : slice.axis_subsets[time_axis_order].interval.low, reverse=reverse)

    return slices


def __get_time_axis_order(axis_subsets):
    """
    Return the time axis order from the list of axis subsets
    :param list subsets: list of analyzed subsets of a slice
    :return: int axis_order: the order of time axis in the list of axis subsets
    """
    i = 0
    for axis_subset in axis_subsets:
        # time axis will be in datetime format, e.g "2001-01-01T01:00:00Z", interval could be decimal value.
        if '"' in str(axis_subset.interval.low):
            return i
        i += 1

    return None


def get_null_values(default_null_values):
    """
    Parse from 1 list of user-defined null values to individual null values
     (e.g: ["20", "30, 40, 50"] -> [20, 30, 40, 50])
    :param list[str] default_null_values: user defined null values in ingredient file
    """
    null_values = None
    if default_null_values is not None:
        null_values = []
        for value in default_null_values:
            values = str(value).strip().split(",")
            values = [x.strip() for x in values]
            null_values += values

    return null_values
